VideoWeb/VideoWeb/ClientApp/src/app/services/user-media-stream.service.ts

```diff
@@ -2,106 +2,238 @@ import { Injectable } from '@angular/core';
+        this.mediaStreamService
+            .getStreamForCam(cameraDevice)
+            .pipe(take(1))
+            .subscribe(cameraStream => {
+                this.logger.debug(`${this.loggerPrefix} Got camera stream`, {
+                    cam: cameraDevice,
+                    stream: cameraStream,
+                    isAudioOnly: this.isAudioOnly
+                });
 
-    async getStreamForMic(device: UserMediaDevice): Promise<MediaStream> {
-        try {
-            if (device) {
-                return await this.navigator.mediaDevices.getUserMedia({ audio: { deviceId: { exact: device.deviceId } } });
-            } else {
-                return this.getDefaultMicStream();
-            }
-        } catch (error) {
-            this.logger.error(`${this.loggerPrefix} Could not get audio stream for microphone`, error);
-            this.errorService.handlePexipError(new CallError(error.name), null);
-        }
+                this.currentStream = this.mediaStreamService.initialiseNewStream(cameraStream.getTracks());
+                this.activateCameraStream = cameraStream;
+
+                this.mediaStreamService
+                    .getStreamForMic(microphoneDevice)
+                    .pipe(take(1))
+                    .subscribe(microphoneStream => {
+                        this.logger.debug(`${this.loggerPrefix} Got microphone stream`, {
+                            mic: microphoneDevice,
+                            stream: microphoneStream,
+                            isAudioOnly: this.isAudioOnly
+                        });
+
+                        microphoneStream.getTracks().forEach(track => {
+                            this.currentStream.addTrack(track);
+                        });
+
+                        this.activateMicrophoneStream = microphoneStream;
+                        this.currentStreamSubject.next(this.currentStream);
+
+                        this.logger.info(`${this.loggerPrefix} Built initial stream`, {
+                            mic: microphoneDevice,
+                            cam: cameraDevice,
+                            stream: this.currentStream,
+                            isAudioOnly: this.isAudioOnly
+                        });
+
+                        this.logger.debug(`${this.loggerPrefix} Subscribing to active video and microphone device changes`);
+                        this.userMediaService.activeVideoDevice$.subscribe(videoDevice => {
+                            this.onActiveCameraChanged(videoDevice);
+                        });
+
+                        this.userMediaService.activeMicrophoneDevice$.subscribe(activeMicrophoneDevice => {
+                            this.onActiveMicrophoneChanged(activeMicrophoneDevice);
+                        });
+
+                        this.userMediaService.isAudioOnly$.subscribe(audioOnly => {
+                            this.onIsAudioOnlyChanged(audioOnly);
+                        });
```

Comment:
Unsubscribe?

Reply:
I think no need unsubscribe because angular service never get destroyed, so there is no chance of memory leaks

Reply:
I have thought a bit about that and what would be best. I will have to look into it.

Reply:
Perhaps if we look to start using scoped services which might make sense for something like this, this would be important but lets not worry about that now. And there doesn't seem to be any risk of duplicates of these happening so happy with that.